File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/64 bits/xorshift64.pas
{
 PRNG .......: XorShift (64 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
}

program XorShift64;

{$mode objfpc}

uses
  SysUtils;

type
  xorShift64State = record
    seed: UInt64;
  end;

function xorShift64(var state: xorShift64State): UInt64;
begin
  state.seed := state.seed xor (state.seed shl 13);
  state.seed := state.seed xor (state.seed shr 7);
  state.seed := state.seed xor (state.seed shl 17);
  xorShift64 := state.seed and UInt64($7FFFFFFFFFFFFFFF);
end;

var
  Xn1, Xn2: xorShift64State;
  i: Integer;
begin
  Xn1.seed := 1234;
  Xn2.seed := UInt64(Trunc(Now * 24 * 60 * 60));

  WriteLn('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  WriteLn('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  WriteLn('PADRAO: XorShift (64 bits)');
  WriteLn('=================================================');
  WriteLn;
  WriteLn('   Semente: 1234          Semente: TIME');
  WriteLn('   --------------------   --------------------');

  for i := 1 to 10 do
    WriteLn(Format('%23d%23d', [Int64(xorShift64(Xn1)), Int64(xorShift64(Xn2))]));

  ReadLn;
end.

File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/64 bits/xorshift64.c
/*
 PRNG .......: XorShift (64 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
*/

#include <stdio.h>
#include <stdint.h>
#include <time.h>

struct xorShift64State {
  uint64_t seed;
};

uint64_t xorshift64(struct xorShift64State *state)
{
  uint64_t random = state->seed;
  random ^= random << 13;
  random ^= random >> 7;
  random ^= random << 17;
  return state->seed = random;
}

int main(void)
{
	
  struct xorShift64State Xn1;
  struct xorShift64State Xn2;

  Xn1.seed = 1234;
  Xn2.seed = time(NULL);

  printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
  printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
  printf("PADRAO: XorShift (64 bits)\n");
  printf("=================================================\n");
  printf("\n");
  printf("   Semente: 1234          Semente: TIME\n");
  printf("   --------------------   --------------------\n");

  for (int i = 1; i <= 10; i++)
    printf("%23llu%23llu\n", xorshift64(&Xn1), xorshift64(&Xn2));
    
  return 0;
}
